### PR TITLE
fix(sdk): unable to capture resources inside immutable data containers

### LINKED ISSUE #1156: Unable to capture an array/map/set of resources
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

I tried to capture an array with two buckets:

```js
bring cloud;

let arr = [new cloud.Bucket(), new cloud.Bucket()];

new cloud.Function(inflight (s: str): str => {
  let b1 = arr.at(0);
  let b2 = arr.at(1);
  b1.put("hello", "world");
});
```

### I expected this:

To work

### Instead, this happened:

The capturing mechanism currently treats data and resources differently, which means that in this case the buckets are not bound as resources.

### Component:

Compiler, SDK

### Wing Version:

_No response_

### Node.js Version:

_No response_

### Platform(s):

_No response_

### Anything else?

See https://github.com/winglang/wing/pull/1144#discussion_r1069527720

### PR DESCRIPTION
**Pull request by eladb**
This is an initial step towards merging `data` and `resources` in the binding system. We now allow any immutable data object to also include resources (e.g. an array of buckets, map of queues, etc).

To support this, we use duck-typing to determine if an object is a resource (e.g. it implements `_toInflight` and `_bind`), and call these methods in their relevant place respectively.

Fixes #1156

*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingsdk/src/core/internal.ts [hunk 1]
```diff
@@ -103,8 +105,14 @@ function serializeImmutableData(obj: any): string {
         return `new Map(${serializeImmutableData(Array.from(obj))})`;
       }
 
-      // structs are just objects
-      if (obj instanceof Object) {
+      // if the object is a resource (i.e. has a "_toInflight" method"), then we lift and add to the
+      // list of bindings.
+      if (typeof (obj as IResource)._toInflight === "function") {
+        return (obj as IResource)._toInflight().text;
+      }
```
Comment (Chriscbr) [inline libs/wingsdk/src/core/internal.ts:112]: This comment does not match this code.

BTW - I feel like it's okay to treat constructs without `_toInflight` or `_bind` as still resources - they're just resources that don't have any inflight functionality. For example we could provide a default `_toInflight()` implementation that returns an empty inflight client class, and a bind method that returns nothing.

File: libs/wingsdk/src/core/resource.ts [hunk 2]
```diff
@@ -353,3 +341,69 @@ export class Display {
...
+        (obj as IResource)._bind(host, ops);
+
+        // add connection metadata (use "?" to indicate that we don't know the operation)
+        for (const op of ops) {
+          Resource.addConnection({
+            from: host,
+            to: obj,
+            relationship: op,
+          });
+        }
```
Comment (Chriscbr) [inline libs/wingsdk/src/core/resource.ts:394]: Update comment

File: libs/wingsdk/src/core/resource.ts [hunk 3]
```diff
@@ -353,3 +341,69 @@ export class Display {
...
+
+        return;
+      }
+
+      // structs are just plain objects
+      if (obj.constructor.name === "Object") {
+        Object.values(obj).forEach((item) => bindObject(item, host, ops));
```
Comment (Chriscbr) [inline libs/wingsdk/src/core/resource.ts:401]: I feel like there's something we're not quite modeling here -- the list of operations passed to this function are specifically operations about `obj`. (Or at least, that was the original intent). If a struct contains a bucket and a key value store, we shouldn't mix up performing `put` on a `cloud.Bucket` with performing `put` on a `cloud.KeyValueStore`. Do you see what I mean?

File: libs/wingsdk/src/core/internal.ts (hunk 1 above)
Comment (eladb) [reply libs/wingsdk/src/core/internal.ts:112]: > BTW - I feel like it's okay to treat constructs without _toInflight or _bind as still resources - they're just resources that don't have any inflight functionality. For example we could provide a default _toInflight() implementation that returns an empty inflight client class, and a bind method that does nothing

That's an interesting idea. There is even a use case for referencing things like resource attributes from inflight.

```js
bring "aws-cdk" as aws;

let bucket = new aws.s3.Bucket();

let handler = inflight () => {
  print(bucket.bucket_name);
};
```

I'll open an issue and follow up.

Comment (eladb) [reply libs/wingsdk/src/core/internal.ts:112]: https://github.com/winglang/wing/issues/1251

File: libs/wingsdk/src/core/resource.ts (hunk 2 above)
Comment (eladb) [reply libs/wingsdk/src/core/resource.ts:394]: DONE

File: libs/wingsdk/src/core/resource.ts (hunk 3 above)
Comment (eladb) [reply libs/wingsdk/src/core/resource.ts:401]: Yes, good point. In the `struct` case, `ops` is just going to be `["?"]` anyway, so I think we are "fine".

As discussed yesterday, we need to revisit how "ops" are determined/propagated anyway.

**Review by Chriscbr — APPROVED** [review]

**Comment by Chriscbr** [comment]
@eladb @MarkMcCulloh I'm not sure why this got automatically merged, it looks like on the last commit before mergify commented, the E2E tests were failing. Is this an issue with our configuration somewhere?

**Comment by MarkMcCulloh** [comment]
@Chriscbr I think I see the issue. The self mutation workflow also has a job named "build" so when that passes it considers it good to merge. In this PR I changed the name https://github.com/winglang/wing/pull/1250

**Comment by Chriscbr** [comment]
Ahh, nice catch 👍

**Pull request merged**